helper/schema/schema.go

```diff
@@ -1431,6 +1465,59 @@ func (m schemaMap) diffString(
 	return nil
 }
 
+// handleDiffSuppressOnRefresh visits each of the attributes set in "new" and,
+// if the corresponding schema sets both DiffSuppressFunc and
+// DiffSuppressOnRefresh, checks whether the new value is materially different
+// than the old and if not it overwrites the new value with the old one,
+// in-place.
+func (m schemaMap) handleDiffSuppressOnRefresh(oldState, newState *terraform.InstanceState) {
+	if newState == nil || oldState == nil {
+		return // nothing to do, then
+	}
+
+	// We'll populate this in the loop below only if we find at least one
+	// attribute which needs this analysis.
+	var d *ResourceData
+
+	oldAttrs := oldState.Attributes
+	newAttrs := newState.Attributes
+	for k, newV := range newAttrs {
+		oldV, ok := oldAttrs[k]
+		if !ok {
+			continue // no old value to compare with
+		}
+		if newV == oldV {
+			continue // no change to test
+		}
+
+		schemaList := addrToSchema(strings.Split(k, "."), m)
+		if len(schemaList) == 0 {
+			continue // no schema? weird, but not our responsibility to handle
+		}
+		schema := schemaList[len(schemaList)-1]
+		if !schema.DiffSuppressOnRefresh || schema.DiffSuppressFunc == nil {
+			continue // not relevant
+		}
+
+		if d == nil {
+			// We populate "d" only on demand, to avoid the cost for most
+			// existing schemas where DiffSuppressAlways won't be set.
+			var err error
+			d, err = m.Data(newState, nil)
+			if err != nil {
+				// Should not happen if we got far enough to be doing this
+				// analysis, but if it does then we'll bail out.
+				log.Printf("[WARN] schemaMap.handleDiffSuppressAlways failed to construct ResourceData: %s", err)
+				return
+			}
+		}
+
+		if schema.DiffSuppressFunc(k, oldV, newV, d) {
+			newState.Attributes[k] = oldV // keep the old value, then
```

Comment:
We should probably include a debug log level message when doing this since it could be potentially unexpected behavior for provider developers who are unfamiliar with this functionality:

```suggestion
			log.Printf("[DEBUG] Keeping %q old value due to DiffSuppressFunc", k)
			newState.Attributes[k] = oldV // keep the old value, then
```

Reply:
I'm happy to include this if it seems like it would help with debugging, though it does make me think of some tradeoffs to weigh:
* I think in many cases once this is turned on for a particular attribute it'll appear _every time_ for that attribute, because a typical reason to activate it would be if the developer knows that a remote API does some additional normalization on read which a typical configuration won't match. For example, Terraform's `jsonencode` function always produces minified JSON, whereas some remote APIs intentionally return pretty-printed JSON. This may therefore appear as noise in the logs, but maybe that's not significant given how much other persistent chatter there is in the logs for most runs.
* Since this can happen only with the new flag set, no provider team should encounter this without first learning about that flag and adding it, though indeed perhaps they would enable it and then use this log to understand whether it worked and what effect it had to enable it.
* We don't have any corresponding debug log when we reset `New` to match `Old` in the `Diff` case, which is performing a very similar operation.

With all of that said, I don't feel strongly either way so I'm happy to do whatever feels right to you. Just wanted to think aloud about the tradeoffs.

Reply:
I did add a log line here, though I did it with the `tfsdklog` package instead since I responded to your earlier comment by wiring through the `ctx`.

For consistency I also retroactively added an equivalent log to the plan-time equivalent of this logic, since it felt confusing to me to have `DiffSuppressFunc` outcomes be announced only in one of the two cases where we now check it. However, that other older codepath doesn't have access to `ctx` so I left that one using `log.Printf` for now, even though the log message text is similar.